### PR TITLE
Best practices for some scheduler functions

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , here's another low-priority review

File: lib/pengine/complex.c [hunk 1]
```diff
@@ -182,10 +190,10 @@ get_meta_attributes(GHashTable * meta_hash, pcmk_resource_t * rsc,
     }
 
     for (xmlAttrPtr a = pcmk__xe_first_attr(rsc->xml); a != NULL; a = a->next) {
-        const char *prop_name = (const char *) a->name;
-        const char *prop_value = pcmk__xml_attr_value(a);
-
-        add_hash_param(meta_hash, prop_name, prop_value);
+        if (a->children != NULL) {
+            dup_attr((gpointer) a->name, (gpointer) a->children->content,
```
Comment (nrwahl2) [inline lib/pengine/complex.c:194]: Why did we decide to use `dup_attr()` here (vs. `pcmk__insert_meta()`)? The choice in `expand_parents_fixed_nvpairs()` is clearer

**Review by nrwahl2 — COMMENTED** [review]
Looks good

File: lib/common/nvpair.c [hunk 2]
```diff
@@ -920,26 +920,31 @@ pcmk__xe_attr_is_true(const xmlNode *node, const char *name)
 
 // Meta-attribute handling
 
+/*!
+ * \brief Get the environment variable equivalent of a meta-attribute name
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/nvpair.c:924]: Yep, the hyphens

File: lib/common/nvpair.c [hunk 3]
```diff
@@ -947,21 +947,26 @@ crm_meta_name(const char *attr_name)
     return env_name;
 }
 
+/*!
+ * \brief Get the value of a meta-attribute
+ *
+ * \param[in] meta       Hash table of meta-attributes
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/nvpair.c:953]: Added a long description

File: lib/common/nvpair.c [hunk 4]
```diff
@@ -947,21 +947,26 @@ crm_meta_name(const char *attr_name)
...
+    if ((meta == NULL) || (attr_name == NULL)) {
         return NULL;
     }
-    key = crm_meta_name(field);
-    if (key) {
-        value = g_hash_table_lookup(hash, key);
-        free(key);
-    }
-
+    key = crm_meta_name(attr_name);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/nvpair.c:967]: might as well

File: lib/common/strings.c [hunk 5]
```diff
@@ -687,6 +687,34 @@ pcmk__strkey_table(GDestroyNotify key_destroy_func,
...
+ */
+void
+pcmk__insert_dup(GHashTable *table, const char *name, const char *value)
+{
+    char *name_copy = NULL;
+    char *value_copy = NULL;
+
+    CRM_ASSERT((table != NULL) && (name != NULL));
+
+    name_copy = strdup(name);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/strings.c:708]: thanks

File: lib/common/strings.c [hunk 6]
```diff
@@ -687,6 +687,34 @@ pcmk__strkey_table(GDestroyNotify key_destroy_func,
...
+    CRM_ASSERT(name_copy != NULL);
+
+    if (value != NULL) {
+        value_copy = strdup(value);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/strings.c:712]: thanks

File: lib/pengine/complex.c (hunk 1 above)
Comment (kgaillot) [reply lib/pengine/complex.c:194]: `pcmk__insert_meta()` inserts into `(obj)->meta`, whereas here we have the table that we need to insert into.

To override or not doesn't matter here as long as we pass an empty hash table to `get_meta_attributes()`. I just noticed there is one caller that does not: `register_if_fencing_device()` unpacks a resource's meta-attributes a second time passing a node for rule evaluation. The evaluated values won't override any existing non-rule-based values. That's probably a bug.

**Comment by kgaillot** [comment]
Updated per review

File: include/crm/common/nvpair_internal.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+/*!
+ * \internal
+ * \brief Insert a meta-attribute into a hash table
+ *
+ * \param[in] obj    Resource (pe_resource_t) or action (pe_action_t) to add to
+ * \param[in] name   Meta-attribute name
+ * \param[in] value  Value to add
+ */
+#define pcmk__insert_meta(obj, name, value) do {                        \
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply include/crm/common/nvpair_internal.h:27]: Lab in progress.

`add_hash_param()` was added early on (ea1359b32), and `dup_attr()` in the very next commit (5050638c6), so I suspect that it was designed for the `dup_attr()` case and (much like `safe_str_eq()` being case-insensitive) it was used elsewhere out of convenience ignoring that aspect. (Even in the original commit, all uses look like they should override.)

**Review by nrwahl2 — APPROVED** [review]

**Pull request merged**